src/main/java/com/mojang/serialization/DynamicOps.java

```diff
@@ -0,0 +1,287 @@
+
+    default DataResult<T> mergeToMap(final T map, final Map<T, T> values) {
+        return mergeToMap(map, MapLike.forMap(values, this));
+    }
+
+    default DataResult<T> mergeToMap(final T map, final MapLike<T> values) {
+        // TODO: AtomicReference.getPlain/setPlain in java9+
+        final MutableObject<DataResult<T>> result = new MutableObject<>(DataResult.success(map));
+
+        values.entries().forEach(entry ->
+            result.setValue(result.getValue().flatMap(r -> mergeToMap(r, entry.getFirst(), entry.getSecond())))
+        );
+        return result.getValue();
+    }
+
+    /**
+     * Only successful if first argument is empty
+     */
+    default DataResult<T> mergeToPrimitive(final T prefix, final T value) {
+        if (!Objects.equals(prefix, empty())) {
+            return DataResult.error("Do not know how to append a primitive value " + value + " to " + prefix, value);
+        }
+        return DataResult.success(value);
+    }
+
+    DataResult<Stream<Pair<T, T>>> getMapValues(T input);
+
+    default DataResult<Consumer<BiConsumer<T, T>>> getMapEntries(final T input) {
+        return getMapValues(input).map(s -> c -> s.forEach(p -> c.accept(p.getFirst(), p.getSecond())));
+    }
+
+    T createMap(Stream<Pair<T, T>> map);
+
+    default DataResult<MapLike<T>> getMap(final T input) {
+        return getMapValues(input).flatMap(s -> {
+            try {
+                return DataResult.success(MapLike.forMap(s.collect(Pair.toMap()), this));
+            } catch (final IllegalStateException e) {
+                return DataResult.error("Error while building map: " + e.getMessage());
+            }
+        });
+    }
+
+    default T createMap(final Map<T, T> map) {
+        return createMap(map.entrySet().stream().map(e -> Pair.of(e.getKey(), e.getValue())));
+    }
+
+    DataResult<Stream<T>> getStream(T input);
+
+    default DataResult<Consumer<Consumer<T>>> getList(final T input) {
+        return getStream(input).map(s -> s::forEach);
+    }
+
+    T createList(Stream<T> input);
+
+    default DataResult<ByteBuffer> getByteBuffer(final T input) {
+        return getStream(input).flatMap(stream -> {
+            final List<T> list = stream.collect(Collectors.toList());
+            if (list.stream().allMatch(element -> getNumberValue(element).result().isPresent())) {
+                final ByteBuffer buffer = ByteBuffer.wrap(new byte[list.size()]);
+                for (int i = 0; i < list.size(); i++) {
+                    buffer.put(i, getNumberValue(list.get(i)).result().get().byteValue());
+                }
+                return DataResult.success(buffer);
+            }
+            return DataResult.error("Some elements are not bytes: " + input);
+        });
+    }
+
+    default T createByteList(final ByteBuffer input) {
+        final int[] i = {0};
+        return createList(Stream.generate(() -> createByte(input.get(i[0]++))).limit(input.capacity()));
```

Comment:
Feels like a weird structure above, any reason this couldn't work?

```suggestion
        return createList(IntStream.range(0, input.capacity()).mapToObj(i -> createByte(input.get(i))));
```

Reply:
:+1: